File: go/data_structure/7_chapter/kmp/kmp.go
package algorithm

func Match(str, pattern []byte) int {
	next := buildNext(pattern)
	m, n := len(str), len(pattern)
	i, j := 0, 0
	for j < n && i < m {
		if j < 0 || str[i] == pattern[j] {
			i++
			j++
		} else {
			j = next[j]
		}
	}
	if j == n {
		return i - j
	} else {
		return -1
	}
}

func buildNext(pattern []byte) []int {
	next := make([]int, len(pattern))
	next[0] = -1
	i, j := 0, -1
	m := len(pattern)
	for i < m-1 {
		if j < 0 || pattern[j] == pattern[i] {
			i++
			j++
			if pattern[i] == pattern[j] {
				next[i] = next[j]
			} else {
				next[i] = j
			}
		} else {
			j = next[j]
		}
	}
	return next
}

File: cpp/data_structures/eleven_chapter/KMP.c
#include <cstring>

int* buildNext(char*);

int match(char* P, char* T) {
	int* next = buildNext(P);
	int n = strlen(T), i = 0;
	int m = strlen(P), j = 0;
	if (!m)
		return 0;
	if (!n || m > n)
		return -1;
	while (j < m && i < n)
		if (j < 0 || P[j] == T[i]) {
			++i;
			++j;
		} else 
			j = next[j];
	delete [] next;
	return i - j > n - m ? - 1 : i - j;
}
int* buildNext(char* P) {
	int m = strlen(P), j = 0;
	int* N = new int[m];
	int t = N[0] = -1;
	while (j < m - 1)
		if (t < 0 || P[j] == P[t]) {
			++j;
			++t;
			N[j] = P[j] == P[t] ? N[t] : t;
		} else
			t = N[t];


	return N;
}
